components/patient-access-rules/api/src/main/java/org/phenotips/data/permissions/script/RecordLockingService.java

```diff
@@ -0,0 +1,115 @@
+import org.xwiki.context.Execution;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.script.service.ScriptService;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * A service to add or remove a lock object to a record. The lock object will remove edit rights for all users on the
+ * document.
+ * @version $Id$
+ * @since 1.3
+ */
+@Component
+@Named("recordLocking")
+@Singleton
+public class RecordLockingService implements ScriptService
+{
+    /** The XClass used for lock objects. */
+    private EntityReference lockClassReference = new EntityReference("PatientLock", EntityType.DOCUMENT,
+        Constants.CODE_SPACE_REFERENCE);
+
+    /** Provides access to the current request context. */
+    @Inject
+    private Execution execution;
+
+    /** Allows retrieval of Patient from ID*/
+    @Inject
+    PatientRepository pr;
+
+    /** Allows checking of access rights on a patient*/
+    @Inject
+    PermissionsManager pm;
+
+    /**
+     * Sets or removes a lock from a record. Locks remove edit rights from all users.
+     * @param patientId The record to be locked
+     * @param lock "true" to create a lock, "false" to remove
+     * @return true if successful false if otherwise
+     */
+    public Boolean setRecordLock (String patientId, Boolean lock)
+    {
+        if (patientId == null || lock == null) {
+            return Boolean.FALSE;
+        }
+
+        XWikiContext context = (XWikiContext) this.execution.getContext().getProperty("xwikicontext");
+        XWiki xwiki = context.getWiki();
+
+        Patient patient = this.pr.getPatientById(patientId);
+        if (patient == null) {
+            return Boolean.FALSE;
+        }
+
+        try {
+            DocumentReference patientDocumentReference = patient.getDocument();
+            XWikiDocument patientDocument = xwiki.getDocument(patientDocumentReference, context);
+            PatientAccess patientAccess = this.pm.getPatientAccess(patient);
+            Boolean hasLockingPermission = patientAccess.hasAccessLevel(new ManageAccessLevel());
+
+            //Edit the lock if the user has locking permission
+            if (hasLockingPermission) {
+                BaseObject previousLockObject = patientDocument.getXObject(lockClassReference);
+
+                if (lock && previousLockObject == null) {
+                    patientDocument.createXObject(lockClassReference, context);
+                } else if (!lock && previousLockObject != null) {
+                    patientDocument.removeXObjects(lockClassReference);
+                }
+
+                xwiki.saveDocument(patientDocument, context);
```

Comment:
Add a comment and set this as a minor edit:

    xwiki.saveDocument(patientDocument, (lock ? "Locked" : "Unlocked") + " patient record", true, context);